### ISSUE TITLE
CI is failing (timing out) on Julia nightly

### ISSUE THREAD

**Issue by DilumAluthge** [mcabbott/Tullio.jl#61]
On the most recent commit to `master` (`58c1fbb`), CI passed on Julia 1.4 and passed on Julia 1.5 but failed (timed out) on Julia nightly. Links to the logs:
- Julia 1.4.2 (passed in 1h 21m 42s): https://github.com/mcabbott/Tullio.jl/runs/1505716118
- Julia 1.5.3 (passed in 52m 41s ): https://github.com/mcabbott/Tullio.jl/runs/1505716124
- Julia nightly (failed - killed for timing out at 6h 0m 13s): https://github.com/mcabbott/Tullio.jl/runs/1505716131

GitHub Actions will kill a job if it runs for more than six hours.

We need to figure out how to make the CI jobs run for less than six hours.

Here is one option: we separate the test suite into "groups". E.g. if currently the test suite looks like this:
```julia
using Tullio
using Test

@test f() == "a"

@test g() == "b"

@test h() == "c"
```

We instead divide the test suite into three groups as such:
```julia
using Tullio
using Test

test_group = get(ENV, "TULLIO_TEST_GROUP", "all")

if test_group in ["all", "1"]
    @test f() == "a"
end

if test_group in ["all", "2"]
    @test g() == "b"
end

if test_group in ["all", "3"]
    @test h() == "c"
end
```

And then we set up the `matrix` in `.github/workflows/ci.yml` file as such:
```yaml
strategy:
  fail-fast: false
  matrix:
    JULIA_NUM_THREADS:
      - '1'
      - '6'
    TULLIO_TEST_GROUP:
      - '1'
      - '2'
      - '3'
    version:
      - '1.4'
      - '1' # automatically expands to the latest stable 1.x release of Julia
    os:
      - ubuntu-latest
    arch:
      - x64
```

@mcabbott @chriselrod Any thoughts?

**Comment by chriselrod** [comment]
I like this suggestion. I saw something similar for a package tested on a server using GitLab runners, great that it seems relatively easy to setup now with GitHub CI. I think it'd be a good idea for some of my own packages as well.

But I think the 1 hour -> 6 hour regression sounds like something to investigate.

I see it prints some times:
```julia
 Test Summary:          | Pass  Broken  Total
parsing all the things |  244       4    248
Test Summary:        | Pass  Broken  Total
tests from Einsum.jl |  684       5    689
[ Info: Basic tests took 106.6 seconds
Test Summary:   | Pass  Total
internal pieces |  408    408
Test Summary:         | Pass  Total
matrix multiplication |   33     33
Test Summary:                  | Pass  Broken  Total
gradients: Tracker + DiffRules |   62       1     63
Test Summary:                    | Pass  Broken  Total
gradients: Tracker + ForwardDiff |   37       1     38
[ Info: Tracker tests took 197.9 seconds
Error: The operation was canceled.
```
But these printed times don't come close to adding up to the 6 hours it took before getting cancelled.

**Comment by mcabbott** [comment]
I can't believe I have a package whose tests take hours, how did this happen? 

But hopefully not really 6 hours. The next set of tests after Tracker is for KernelAbstractions, so perhaps this timeout is the same as https://github.com/JuliaGPU/KernelAbstractions.jl/issues/155?

Even without timeouts, though, this split is an interesting idea. It would give you quicker feedback on which things are broken. And the implementation actually looks very simple.

**Comment by DilumAluthge** [comment]
Oh, this might be the same as https://github.com/JuliaGPU/KernelAbstractions.jl/issues/155!

Maybe this is the easiest first step: we create two test groups: group 1 is everything except the KernelAbstractions tests, and group 2 is only the KernelAbstractions tests.

**Comment by mcabbott** [comment]
I think this can be closed, thanks again for #62.
